closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurityConfigurator.java

```diff
@@ -0,0 +1,325 @@
+        }
+    }
+
+    /**
+     * Add restricted security providers.
+     *
+     * @param providers the provider name array
+     */
+    private static void setProviders(Properties props, ArrayList<String> providers) {
+
+        if (debug != null) {
+            debug.println("Adding restricted security provider.");
+        }
+
+        int pNum = 1;
+        for (String provider : providers) {
+            props.setProperty("security.provider." + pNum, provider);
+            pNum ++;
+            if (debug != null) {
+                debug.println("Added restricted security provider: " + provider);
+            }
+        }
+    }
+
+    /**
+     * Add restricted security properties.
+     *
+     * @param props the java.security properties
+     */
+    private static void setProperties(Properties props, RestrictedSecurityProperties properties) {
+
+        if (debug != null) {
+            debug.println("Adding restricted security properties.");
+        }
+
+        Map<String, String> propsMapping = new HashMap<>();
+
+        // JDK properties name as Key, restricted security properties vaule as value.
+        propsMapping.put("jdk.tls.disabledNamedCurves", properties.getJdkTlsDisabledNamedCurves());
+        propsMapping.put("jdk.tls.disabledAlgorithms", properties.getJdkTlsDisabledAlgorithms());
+        propsMapping.put("jdk.tls.ephemeralDHKeySize", properties.getJdkTlsDphemeralDHKeySize());
+        propsMapping.put("jdk.tls.legacyAlgorithms", properties.getJdkTlsLegacyAlgorithms());
+        propsMapping.put("jdk.certpath.disabledAlgorithms", properties.getJdkCertpathDisabledAlgorithms());
+        propsMapping.put("jdk.security.legacyAlgorithm", properties.getJdkSecurityLegacyAlgorithm());
+
+        for (Map.Entry<String, String> entry : propsMapping.entrySet()) {
+            String jdkPropsName = entry.getKey();
+            String propsNewValue = entry.getValue();
+
+            String propsOldValue = !isNullOrBlank(props.getProperty(jdkPropsName)) ? props.getProperty(jdkPropsName) : "";
+
+            if (!isNullOrBlank(propsNewValue)) {
+                String values = !isNullOrBlank(propsOldValue) ? propsOldValue + ", " + propsNewValue : propsNewValue;
+                props.setProperty(jdkPropsName, values);
+                if (debug != null) {
+                    debug.println("Added restricted security properties, with property: " + jdkPropsName + " values: "
+                            + values);
+                }
+            }
+        }
+
+        // For keyStore and keystore.type, old value not needed, just set the new value.
+        if (!isNullOrBlank(properties.getKeyStoreType())) {
+            props.setProperty("keystore.type", properties.getKeyStoreType());
+        }
+        if (!isNullOrBlank(properties.getKeyStore())) {
+            System.setProperty("javax.net.ssl.keyStore", properties.getKeyStore());
```

Comment:
Common subexpressions `properties.getKeyStoreType()`, `properties.getKeyStore()`.
Why does this use `System` instead of `props` on line 287?

Reply:
@keithc-ca 

"keystore.type" is a property defined in the java.security file, so we can use props.setProperty() to set its value.

"javax.net.ssl.keyStore" is a SSL properties set at the JVM level via system properties. So, we have to either set them when running the program (java -D....), or set them in code by doing System.setProperty().

Reply:
@keithc-ca 

"keystore.type" is a Java security property defined in the Java security properties file(the java.security file). The Security.java loads this properties file as a Properties object. so we can use Properties.setProperty() to set or change its value.

"javax.net.ssl.keyStore" is a SSL property, means the location of the Java keyStore file containing an application its own certificates and private keys. In the FIPS mode, it using the NSS database(when keystore.type set to PKCS11 format) to store the certificates and keys, so the keyStore file location must be set to NONE. That mean, we must set "javax.net.ssl.keyStore" to NONE. And due to the SSL properties set at the JVM level via system properties, not read by the codes as a Properties object. So, we have to either set them when running the program (java -D....), or set them in code by doing System.setProperty().